File: sections/24_grid_dp/problems/b_lowest_toll/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    it = iter(data)
    n = next(it)
    m = next(it)
    inf = 10**30
    prev = [inf] * m
    for r in range(n):
        cur = [inf] * m
        for c in range(m):
            x = next(it)
            if r == 0 and c == 0:
                cur[c] = x
            else:
                best = prev[c]
                if c:
                    best = min(best, cur[c - 1])
                cur[c] = best + x
        prev = cur
    print(prev[-1])


if __name__ == "__main__":
    main()

File: sections/24_grid_dp/problems/b_lowest_toll/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    if (!(cin >> n >> m)) return 0;
    const long long INF = (1LL << 62);
    vector<long long> prev(m, INF), cur(m, INF);
    for (int r = 0; r < n; r++) {
        fill(cur.begin(), cur.end(), INF);
        for (int c = 0; c < m; c++) {
            long long x;
            cin >> x;
            if (r == 0 && c == 0) {
                cur[c] = x;
            } else {
                long long best = prev[c];
                if (c) best = min(best, cur[c - 1]);
                cur[c] = best + x;
            }
        }
        swap(prev, cur);
    }
    cout << prev[m - 1] << '\n';
    return 0;
}

